src/main/java/uk/gov/hmcts/reform/sandl/snlevents/actions/hearing/WithdrawHearingAction.java

```diff
@@ -0,0 +1,154 @@
+        this.hearingPartRepository = hearingPartRepository;
+        this.statusConfigService = statusConfigService;
+        this.statusServiceManager = statusServiceManager;
+        this.objectMapper = objectMapper;
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public void getAndValidateEntities() {
+        hearing = hearingRepository.findOne(withdrawHearingRequest.getHearingId());
+        hearingParts = hearing.getHearingParts();
+        sessions = hearingParts.stream()
+            .map(HearingPart::getSession)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        // Validation moved to act() due to conflict with optimistic lock
+
+        if (!statusServiceManager.canBeWithdrawn(hearing)) {
+            throw new SnlEventsException("Hearing can not be withdrawn");
+        }
+    }
+
+    @Override
+    public UUID[] getAssociatedEntitiesIds() {
+        val ids = hearing.getHearingParts().stream().map(HearingPart::getId).collect(Collectors.toList());
+        ids.addAll(hearing.getHearingParts().stream()
+            .map(HearingPart::getSessionId)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList()));
+        ids.add(withdrawHearingRequest.getHearingId());
+
+        return ids.stream().toArray(UUID[]::new);
+    }
+
+    @Override
+    public void act() {
+        try {
+            previousHearing = objectMapper.writeValueAsString(hearing);
+        } catch (JsonProcessingException e) {
+            throw new SnlRuntimeException(e);
+        }
+
+        hearing.setStatus(statusConfigService.getStatusConfig(Status.Withdrawn));
+        entityManager.detach(hearing);
+        hearing.setVersion(withdrawHearingRequest.getHearingVersion());
+        hearingRepository.save(hearing);
+
+        originalHearingParts = utdps.mapHearingPartsToStrings(objectMapper, hearingParts);
+        hearingParts.stream().forEach(hp -> {
+            hp.setSession(null);
+            hp.setSessionId(null);
+            if (hp.getStatus().getStatus() == Status.Listed) {
+                hp.setStatus(statusConfigService.getStatusConfig(Status.Vacated));
+            } else if (hp.getStatus().getStatus() == Status.Unlisted) {
+                hp.setStatus(statusConfigService.getStatusConfig(Status.Withdrawn));
+            }
+        });
+
+        hearingPartRepository.save(hearingParts);
+    }
+
+    @Override
+    public List<UserTransactionData> generateUserTransactionData() {
+        originalHearingParts.forEach((id, hpString) ->
+            utdps.prepareUserTransactionDataForUpdate("hearingPart", id, hpString,  0)
+        );
+
+        utdps.prepareUserTransactionDataForUpdate("hearing", hearing.getId(),
+            previousHearing, 1);
+
+        sessions.stream().forEach(s ->
+            utdps.prepareLockedEntityTransactionData("session", s.getId(), 0)
+        );
+
+        return utdps.getUserTransactionDataList();
```

Comment:
These preparers are really nice. Would be even better to create a UserTransactionDataBuilder and use it like:

List<UserTransactionData> utdList = new UserTransactionDataBuilder()
     .withHearingPartUpdate(id, hpString)
     .withHearingUpdate(id, hString)
     .withSessionLock(id)
     .build();

But that's just a loose idea, I don't want you to implement it, it's great the way it is